src/main/java/com/google/devtools/build/lib/remote/blobstore/http/HttpBlobStore.java

```diff
@@ -309,13 +328,34 @@ private void releaseDownloadChannel(Channel ch) {
     if (ch.isOpen()) {
       // The channel might have been closed due to an error, in which case its pipeline
       // has already been cleared. Closed channels can't be reused.
-      ch.pipeline().remove(ReadTimeoutHandler.class);
-      ch.pipeline().remove(HttpClientCodec.class);
-      ch.pipeline().remove(HttpDownloadHandler.class);
+      try {
+        ch.pipeline().remove(ReadTimeoutHandler.class);
+        ch.pipeline().remove(HttpClientCodec.class);
+        ch.pipeline().remove(HttpDownloadHandler.class);
+      } catch (NoSuchElementException e) {
+        // If the channel is in the process of closing but not yet closed, some handlers could have
+        // been removed and would cause NoSuchElement exceptions to be thrown. Because handlers are
+        // removed in reverse-order, if we get a NoSuchElement exception, the following handlers
+        // should have been removed.
+      }
     }
     channelPool.release(ch);
   }
 
+  private void clearChannelPipeline(ChannelPipeline pipeline) {
```

Comment:
I think we don't want to clear the pipeline but actually fail if there's anything in there.
It would be a bug in our code if the pipeline were not empty.

Reply:
Okay, I've updated the two `acquire*Channel()` methods to check if the pipeline is not empty and fail the promise if so. The pipeline could be in a strange state if there are any handlers on a newly acquired channel, so you're right that failing probably makes sense there.